src/day23.scala

```diff
@@ -0,0 +1,158 @@
+import scala.collection.mutable
+
+
+@main def part1(): Unit =
+  val answer = part1(readInput())
+  println(s"The answer is: $answer")
+
+@main def part2(): Unit =
+  val answer = part2(readInput())
+  println(s"The answer is: $answer")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day23"))(_.mkString)
+
+type Energy = Int
+
+enum Room(val x: Int):
+  case A extends Room(3)
+  case B extends Room(5)
+  case C extends Room(7)
+  case D extends Room(9)
+
+enum Amphipod(val energy: Energy, val destination: Room):
+  case A extends Amphipod(1, Room.A)
+  case B extends Amphipod(10, Room.B)
+  case C extends Amphipod(100, Room.C)
+  case D extends Amphipod(1000, Room.D)
+
+object Amphipod:
+  def tryParse(input: Char): Option[Amphipod] =
+    input match
+      case 'A' => Some(Amphipod.A)
+      case 'B' => Some(Amphipod.B)
+      case 'C' => Some(Amphipod.C)
+      case 'D' => Some(Amphipod.D)
+      case _ => None
+
+case class Position(val x: Int, val y: Int)
+
+val hallwayStops: Seq[Position] = Seq(
+  Position(1, 1),
+  Position(2, 1),
+  Position(4, 1),
+  Position(6, 1),
+  Position(8, 1),
+  Position(10, 1),
+  Position(11, 1)
+)
+
+case class Situation(positions: Map[Position, Amphipod], roomSize: Int):
+  def moveAllAmphipodOnce: Seq[(Situation, Energy)] =
+    for
+      (start, amphipod) <- positions.toSeq
+      stop <- nextStops(amphipod, start)
+      path = getPath(start, stop)
+      if path.forall(isEmpty)
+    yield
+      val newPositions = positions - start + (stop -> amphipod)
+      val energy = path.size * amphipod.energy
+      (copy(positions = newPositions), energy)
+
+  def isFinal =
+    positions.forall((position, amphipod) => position.x == amphipod.destination.x)
+
+  /**
+   * Return a list of positions to which an amphipod at position `from` can go
+   * If the amphipod is in its destination room and the room is free it must not go anywhere
+   * If the amphipod is in its destination room and the room is not free it can go to the hallway
+   * If the amphipod is in the hallway it can only go its destination.
+   * Otherwise it can go to the hallway.
+   */
+  private def nextStops(amphipod: Amphipod, from: Position): Seq[Position] =
+    from match
+      case Position(x, y) if x == amphipod.destination.x =>
+        if isDestinationFree(amphipod) then Seq.empty
+        else hallwayStops
+      case Position(_, 1) => getDestination(amphipod).toSeq
+      case _ => hallwayStops
+  
+  /**
+   * Get the destination position of an amphipod in the current situation
+   * It returns `None` if the destination room is not free.
+   * It returns `Some(position)` if the destination room is free where `position` is the lowest
+   * empty position of the room.
+   * It returns `None` if the destination room is full.
+   */
+  private def getDestination(amphipod: Amphipod): Option[Position] =
+    if isDestinationFree(amphipod) then
+      (roomSize + 1).to(2, step = - 1)
```

Comment:
```suggestion
      (roomSize + 1).to(2, step = -1)
```
I've never seen a coding style where we put a space between a unary operator and its argument.